eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/ForwardHandler.scala

```diff
@@ -24,5 +24,7 @@ import akka.event.DiagnosticLoggingAdapter
  * Simple handler that forwards all messages to an actor
  */
 class ForwardHandler(actor: ActorRef) extends ReceiveHandler {
-  override def handle(implicit ctx: ActorContext, log: DiagnosticLoggingAdapter): Receive = { case msg => actor forward msg}
+  override def handle(implicit ctx: ActorContext, log: DiagnosticLoggingAdapter): Receive = {
+    case msg if !msg.isInstanceOf[ReceiveHandler] => actor forward msg
```

Comment:
Interesting, my first reaction was that this could surely be done with regular pattern matching on type, but it doesn't seem possible.

Reply:
but... why do we need this in the first place? ❓ 😕 🤷‍♂️

Reply:
We're intercepting HTLCs in some integration tests in order to fetch the channel state right before the HTLC is relayed/fulfilled/failed to get some commitment-specific data...for the force-close tests (especially with revoked commitments) it's quite handy!

Reply:
> We're intercepting HTLCs in some integration tests

I got that but why do we need to catch this particular case? Didn't we catch htlcs in the previous version of the tests, without needing this hack?

Reply:
It worked before because we used one eclair node per close test, so we only set a `ForwardHandler` once on each node.

Now that I removed some of the test eclair nodes (F4, F5, etc) and we re-use the same nodes across multiple tests, we need to be able to set a new `ForwardHandler`, which was simply ignored without this line. Does that make sense? An alternative would be to share the `ForwardHandler` between tests, but it feels cleaner to give the possibility to change the `ReceiveHandler` at runtime.

Reply:
Gotcha, how about this then:
```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/PaymentHandler.scala b/eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/PaymentHandler.scala
index a8784d26a..03cf21756 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/PaymentHandler.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/PaymentHandler.scala
@@ -36,13 +36,15 @@ class PaymentHandler(nodeParams: NodeParams, register: ActorRef) extends Actor w

   override def receive: Receive = normal(defaultHandler.handle(context, log))

-  def normal(handle: Receive): Receive = handle orElse {
+  def addReceiveHandler(handle: Receive): Receive = {
     case handler: ReceiveHandler =>
       log.info(s"registering handler of type=${handler.getClass.getSimpleName}")
       // NB: the last handler that was added will be the first called
       context become normal(handler.handle(context, log) orElse handle)
   }

+  def normal(handle: Receive): Receive = addReceiveHandler(handle) orElse handle
+
   override def mdc(currentMessage: Any): MDC = Logs.mdc(category_opt = Some(Logs.LogCategory.PAYMENT))
 }
```

Makes your change to the `ForwardHandler`  unnecessary.

Reply:
Good idea, I'll apply that!

Reply:
A comment seems appropriate:

```scala
  /**
   * This is a bit subtle because we want handlers to be as generic as possible, but we also want to catch a particular
   * type of message (the [[ReceiveHandler]]s themselves) to update the list of handlers.
   *
   * That why we *prepend* new handlers, but after the first special handler [[addReceiveHandler]]:
   * {{{
   *    paymentHandler ! handler1
   *    paymentHandler ! handler2
   *    paymentHandler ! handler3
   *    // the current handler is now addReceiveHandler :: handler3 :: handler2 :: handler1
   * }}}
   *
   * @param handle
   * @return
   */
  def normal(handle: Receive): Receive = addReceiveHandler(handle) orElse handle
```

Reply:
Done in https://github.com/ACINQ/eclair/pull/1494/commits/e65f5a5380f409543f5c98dad1e6bed745bb51f9